state/indexer/sink/psql/backport_test.go

```diff
@@ -0,0 +1,11 @@
+package psql
+
+import (
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/state/txindex"
+)
+
+var (
+	_ indexer.BlockIndexer = BackportBlockIndexer{}
+	_ txindex.TxIndexer    = BackportTxIndexer{}
```

Comment:
Why make this be in a test package? We have tended to put these kinds of assertions in the main package, so that it breaks "main compile"? 

(this kind of thing is one of the reasons I like having functions returning interface types, to avoid misplacing these directives.)

Reply:
I mainly did this because that's the convention I'm used to. 🙂 

There are a couple good reasons to keep satisfaction checks in the test: 

- It makes the production build faster (by removing an unnecessary dependency). 
- Relatedly: since we run tests before merging anyway, slowing down the production build doesn't really buy us a lot. 
- It means we don't _need_ to return interfaces in the production package.

I generally think it's a mistake to return interfaces in a public API, except where it's required for structural reasons (e.g., you have multiple implementations you provide dynamically) or where it's required to satisfy some other interface (e.g., your method signature references another interface). Otherwise, it just complicates the build graph for no clear value.

More importantly: You should rarely if ever return an interface belonging to _another package_, since that means any change to that interface can potentially break your tests even if there is no effect.

Reply:
Yeah, I agree. I think it's fine. :) We definitely put these checks in the main package near the type definition everywhere else. 

I think this whole nexus of conversation is one of those things where the differences that we're talking about are *really* narrow (particularly in our case) like, the amount of compile time about where this lives is minimal either way, and I think interface changes are likely to cause things to break at some point (which I think is a good thing! and means we can rely on the compiler,) and I think my preference is to keep things *as close* as possible to the code you'd need to change to unbreak things? But I think this is just a personal preference, and like I said probably not material to this PR.

Reply:
Agreed that the difference is narrow, and that the "savings" in this case is minor. While these questions aren't all directly related to this PR, I think some of them are worth pulling apart. I wrote them up here: https://gist.github.com/creachadair/d574a92c0feb11615f3ddfa04e08ae95 if you are interested.